src/librustc_typeck/check/demand.rs

```diff
@@ -214,10 +216,12 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                             self_ty.starts_with("std::option::Option") ||
                             self_ty.starts_with("std::result::Result")
                         ) && (name == "map" || name == "and_then");
-                        if is_as_ref_able {
-                            return Some((span.shrink_to_lo(),
-                                         "consider using `as_ref` instead",
-                                         "as_ref().".into()));
+                        match (is_as_ref_able, cm.span_to_snippet(*span)) {
+                            (true, Ok(src)) => {
+                                return Some((span.shrink_to_lo(), "consider using `as_ref` instead",
+                                             format!("as_ref().{}", src)));
```

Comment:
I understand why you are looking to change the suggestion to make it clearer, but this change would cause the code `opt.and_then(|arg| Some(takes_ref(arg)));` to be replaced with `opt.map.and_thenand_then(|arg| Some(takes_ref(arg)));` because the span only points at the point after the dot (containing nothing).

I see two alternatives: change the span to make the new suggestion work, or make the suggestion text longer than 10 words, which is the arbitrary length to inline suggestions instead of showing them on their own, which would make this suggestion easier to follow:

```
LL |   opt.map(|arg| takes_ref(arg));
   |                           ^^^ expected &Foo, found struct `Foo`
   |
help: consider using `as_ref` in order to hold a borrow of the inner element
   |
LL |   opt.as_ref().map(|arg| takes_ref(arg));
   |       ^^^^^^^^^
```

Reply:
Right, the code to be inserted is `as_ref().map`, but the span points at only the space between `.` and `map(`. When inserting the code, we end up with `as_ref().mapmap(`. We need to change the span to include `.map`.

Reply:
Ah! I think see what you're getting at. I changed the suggested span.